frankenstein/chunk_store.go

```diff
@@ -344,125 +369,234 @@ func (c *AWSChunkStore) Get(ctx context.Context, from, through model.Time, match
 		case err := <-incomingErrors:
-			errors = append(errors, err)
+			lastErr = err
 		}
 	}
-	if len(errors) > 0 {
-		return nil, errors[0]
-	}
-	return chunkSets, nil
+	return chunks, lastErr
+}
+
+func next(s string) string {
+	// TODO deal with overflows
+	l := len(s)
+	result := s[:l-1] + string(s[l-1]+1)
+	return result
 }
 
-func (c *AWSChunkStore) lookupChunksFor(userID string, hour int64, metricName model.LabelValue, matchers []*metric.LabelMatcher, incomingChunkSets chan []wire.Chunk, incomingErrors chan error) {
-	var chunkSets [][]wire.Chunk
+func (c *AWSChunkStore) lookupChunksFor(userID string, hour int64, metricName model.LabelValue, matchers []*metric.LabelMatcher) (wire.ChunksByID, error) {
+	if len(matchers) == 0 {
+		return c.lookupChunkForMeticName(userID, hour, metricName)
+	}
+
+	incomingChunkSets := make(chan wire.ChunksByID)
+	incomingErrors := make(chan error)
 	for _, matcher := range matchers {
-		// TODO build support for other matchers
-		if matcher.Type != metric.Equal {
-			incomingErrors <- fmt.Errorf("%s matcher not supported yet", matcher.Type)
-			return
+		go func(matcher *metric.LabelMatcher) {
+			incoming, err := c.lookupChunksForMatcher(userID, hour, metricName, matcher)
+			if err != nil {
+				incomingErrors <- err
+			} else {
+				incomingChunkSets <- incoming
+			}
+		}(matcher)
+	}
+
+	var chunkSets []wire.ChunksByID
+	var lastErr error
+	for i := 0; i < len(matchers); i++ {
+		select {
+		case incoming := <-incomingChunkSets:
+			chunkSets = append(chunkSets, incoming)
+		case err := <-incomingErrors:
+			lastErr = err
 		}
 
-		hashValue := hashValue(userID, hour, metricName)
-		rangeMinValue := rangeValue(matcher.Name, matcher.Value, minChunkID)
-		rangeMaxValue := rangeValue(matcher.Name, matcher.Value, maxChunkID)
+	}
+	return nWayIntersect(chunkSets), lastErr
+}
 
-		var resp *dynamodb.QueryOutput
-		err := timeRequest("Query", dynamoRequestDuration, func() error {
-			var err error
-			resp, err = c.dynamodb.Query(&dynamodb.QueryInput{
-				TableName: aws.String(c.tableName),
-				KeyConditions: map[string]*dynamodb.Condition{
-					hashKey: {
-						AttributeValueList: []*dynamodb.AttributeValue{
-							{S: aws.String(hashValue)},
-						},
-						ComparisonOperator: aws.String("EQ"),
-					},
-					rangeKey: {
-						AttributeValueList: []*dynamodb.AttributeValue{
-							{S: aws.String(rangeMinValue)},
-							{S: aws.String(rangeMaxValue)},
-						},
-						ComparisonOperator: aws.String("BETWEEN"),
+func (c *AWSChunkStore) lookupChunkForMeticName(userID string, hour int64, metricName model.LabelValue) (wire.ChunksByID, error) {
+	hashValue := hashValue(userID, hour, metricName)
+
+	var resp *dynamodb.QueryOutput
+	err := timeRequest("Query", dynamoRequestDuration, func() error {
+		var err error
+		resp, err = c.dynamodb.Query(&dynamodb.QueryInput{
+			TableName: aws.String(c.tableName),
+			KeyConditions: map[string]*dynamodb.Condition{
+				hashKey: {
```

Comment:
When you omit the range key selector, it just returns all the chunks in the hash bucket?